### PR TITLE
Packaging: Remove classpath ordering hack

### PR DESCRIPTION
**Pull request by rjernst**
After the removal of the joda time hack we used to have, we can cleanup
the codebase handling in security, jarhell and plugins to be more picky
about uniqueness. This was originally in #18959 which was never merged.

closes #18959

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/bootstrap/JarHellTests.java [hunk 1]
```diff
@@ -114,7 +112,7 @@ public void testDirAndJar() throws Exception {
     public void testWithinSingleJar() throws Exception {
         // the java api for zip file does not allow creating duplicate entries (good!) so
         // this bogus jar had to be constructed with ant
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/bootstrap/JarHellTests.java]: This comment is out of date; we produce these with https://github.com/jasontedor/duplicate-classes/ now (the previous mechanism for building the bogus jars produced invalid jars which blew up animal-sniffer).

File: core/src/main/java/org/elasticsearch/bootstrap/JarHell.java [hunk 2]
```diff
@@ -198,7 +203,7 @@ public static void checkJarHell(URL urls[]) throws URISyntaxException, IOExcepti
                     public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
                         String entry = root.relativize(file).toString();
                         if (entry.endsWith(".class")) {
-                            // normalize with the os separator
+                            // normalize with the os separator, remove '.class'
                             entry = entry.replace(sep, ".").substring(0,  entry.length() - 6);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/bootstrap/JarHell.java]: I wonder if replacing `entry.length() - 6` with `entry.length() - ".class".length()` would be clearer?

**Review by jasontedor — APPROVED** [review]
LGTM. I left two minor comments, treat them as you wish.

**Comment by rjernst** [comment]
@jasontedor I addressed your comments before merging. Thanks for the review.

**Pull request merged**